gnovm/pkg/gnolang/store.go

```diff
@@ -73,53 +73,276 @@ type Store interface {
 	LogSwitchRealm(rlmpath string) // to mark change of realm boundaries
 	ClearCache()
 	Print()
+}
+
+// TransactionStore is a store where the operations modifying the underlying store's
+// caches are temporarily held in a buffer, and then executed together after
+// executing Write.
+type TransactionStore interface {
+	Store
+
+	// Write commits the current buffered transaction data to the underlying store.
+	// It also clears the current buffer of the transaction.
 	Write()
-	Flush()
 }
 
-// Used to keep track of in-mem objects during tx.
 type defaultStore struct {
-	alloc            *Allocator    // for accounting for cached items
-	pkgGetter        PackageGetter // non-realm packages
-	cacheObjects     map[ObjectID]Object
-	cacheTypes       map[TypeID]Type
-	cacheNodes       map[Location]BlockNode
-	cacheNativeTypes map[reflect.Type]Type // go spec: reflect.Type are comparable
-	baseStore        store.Store           // for objects, types, nodes
-	iavlStore        store.Store           // for escaped object hashes
+	// underlying stores used to keep data
+	baseStore store.Store // for objects, types, nodes
+	iavlStore store.Store // for escaped object hashes
+
+	// transaction-scoped
+	cacheObjects map[ObjectID]Object          // this is a real cache, reset with every transaction.
+	cacheTypes   hashMap[TypeID, Type]        // this re-uses the parent store's.
+	cacheNodes   hashMap[Location, BlockNode] // until BlockNode persistence is implemented, this is an actual store.
+	alloc        *Allocator                   // for accounting for cached items
+
+	// store configuration; cannot be modified in a transaction
+	pkgGetter        PackageGetter         // non-realm packages
+	cacheNativeTypes map[reflect.Type]Type // reflect doc: reflect.Type are comparable
 	pkgInjector      PackageInjector       // for injecting natives
 	nativeStore      NativeStore           // for injecting natives
 	go2gnoStrict     bool                  // if true, native->gno type conversion must be registered.
 
 	// transient
-	opslog  []StoreOp // for debugging and testing.
 	current []string  // for detecting import cycles.
+	opslog  []StoreOp // for debugging and testing.
+}
+
+type hashMap[K comparable, V any] interface {
```

Comment:
I heavily second what @deelawn is saying, this is insanely overcomplicated

Reply:
Well, it'd have to be copy-pasted exactly for cacheTypes and cacheNodes; essentially what the go compiler is doing under the hood to unwind the generics.

What do you find hard to read? The interface or its implementations?

Reply:
It's the implementation. I'm not a huge fan of using generics to begin with, but I admit they have their use cases.  I have no problem with defining a function that accepts some generic type, but this code is using generics to define an interface, then uses them to create the objects that implement the interface, and the implementations have fields that use other generic types. It all seems very elaborate just to be used for `cacheTypes` and `cacheNodes`.

It's when I see something like this that I have to start flexing my brain muscles to try to figure out what is going on:
```go
func newTxLog[K comparable, V any](source hashMap[K, V]) *txLogMap[K, V] {
	return &txLogMap[K, V]{
		source: source,
		dirty:  make(map[K]deletable[V]),
	}
}
``` 
But maybe the problem is me 🤷‍♂️ 😆

Reply:
Check out the updated code after the latest commits. I've refactored hashMap and txLog into its own package, `internal/txlog`. This way we have a small package with a small API. Hopefully this works to make it at least "compartimentalized" rather than being jumbled up with the store code?

Reply:
Yeah, I think that's an acceptable compromise. Thanks 👍